File: 04. linkedList/MyLinkedList.java
class MyLinkedList{
    class ListNode{
        public int val = 0;
        public ListNode next = null;
        public ListNode(){}

        public ListNode(int val) {
            this.val = val;
        }
    }

    private ListNode dummy;
    private ListNode tail;
    private int size;

    //初始化链表
    public MyLinkedList() {
        dummy = new ListNode();
        tail = dummy;
        size = 0;
    }

    //将值为 val 的结点追加到链表尾部
    public void addAtTail(int val) {
        ListNode node = new ListNode(val);
        tail.next = node;
        tail = node;
        size++;
    }

    //插入值val的新结点，使它成为链表的第一个结点
    public void addAtHead(int val) {
        ListNode node = new ListNode(val);
        node.next = dummy.next;
        dummy.next = node;
        if (tail == dummy) {
            tail = node;
        }
        size++;
    }

    // 获得index索引的前一节点
    private ListNode getPrevNode(int index) {
        ListNode front = dummy.next;
        ListNode back = dummy;
        for (int i = 0; i < index && front != null; i++) {
            back = front;
            front = fornt.next;
        }
    }

    //获取链表中第index个结点的值。如果索引无效，则返回-1
    public int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        return getPrevNode(index).next.val;
    }

    // 在链表中的第 index 个结点之前添加值为 val  的结点。
    // 1. 如果 index 等于链表的长度，则该结点将附加到链表的末尾。
    // 2. 如果 index 大于链表长度，则不会插入结点。
    // 3. 如果index小于0，则在头
    public void addAtIndex(int index, int val) {
        if (index > size) {
            return;
        } else if (index == size) {
            addAtTail(val);
        } else if (index < 0) {
            addAtHead(val);
        } else {
            ListNode prev = getPrevNode(index);
            ListNode node = new ListNode(val);
            node.next = prev.next;
            prev.next = node;
            size++;
        }
    }

    //如果索引index有效，则删除链表中的第index个结点
    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        ListNode prev = getPrevNode(index);
        if (tail == prev.next) {
            tail = prev;
        }
        prev.next = prev.next.next;
        size--;
    }
}

File: 04. linkedList/MyLinkedList.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class MyLinkedList:
    def __init__(self):
        self.dummy = ListNode()
        self.tail = self.dummy
        self.length = 0

    def add_at_tail(self, val):
        self.tail.next = ListNode(val)
        self.tail = self.tail.next
        self.length += 1

    def add_at_head(self, val):
        cur_node = ListNode(val)
        cur_node.next = self.dummy.next
        self.dummy.next = cur_node

        if self.dummy == self.tail:
            self.tail = cur_node
        self.length += 1

    def get(self, index):
        pre_node = self._get_pre_node(index)
        return pre_node.next.val

    def add_at_index(self, index, val):
        if index > self.length:
            return
        elif index == self.length:
            self.add_at_tail(val)
        elif index <= 0:
            self.add_at_head(val)
        else:
            node_pre = self._get_pre_node(index)
            node_cur = ListNode(val, node_pre.next)
            node_pre.next = node_cur
        self.length += 1

    def delete_at_index(self, index):
        if index < 0 or index >= self.length:
            return

        node_pre = self._get_pre_node(index)

        if node_pre.next == self.tail:
            self.tail = node_pre

        node_pre.next = node_pre.next.next
        self.length -= 1

    def _get_pre_node(self, index):
        front = self.dummy.next
        back = self.dummy
        i = 0
        while i < index and front is not None:
            back = front
            front = front.next
            i += 1

        return back